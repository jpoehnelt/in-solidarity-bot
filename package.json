{
  "name": "in-solidarity-bot",
  "version": "1.0.0",
  "private": true,
  "description": "A Probot app",
  "keywords": [
    "probot",
    "github",
    "probot-app"
  ],
  "homepage": "https://github.com/jpoehnelt/in-solidarity-bot",
  "bugs": "https://github.com/jpoehnelt/in-solidarity-bot/issues",
  "repository": "https://github.com/jpoehnelt/in-solidarity-bot.git",
  "license": "Apache 2.0",
  "author": "Justin Poehnelt <justin.poehnelt@gmail.com>",
  "scripts": {
    "build": "tsc -p tsconfig.json",
    "format": "eslint src/*.ts --fix",
    "lint": "eslint src/*.ts",
    "start": "probot run ./dist/index.js",
    "test": "jest --passWithNoTests"
  },
  "jest": {
    "testEnvironment": "node"
  },
  "dependencies": {
    "gitdiff-parser": "^0.2.2",
    "probot": "^9.13.0"
  },
  "devDependencies": {
    "@rollup/plugin-node-resolve": "^8.4.0",
    "@types/jest": "^24.0.19",
    "@types/nock": "^11.1.0",
    "@types/node": "^14.0.27",
    "@typescript-eslint/eslint-plugin": "^3.8.0",
    "@typescript-eslint/parser": "^3.8.0",
    "eslint": "^7.6.0",
    "eslint-config-prettier": "^6.11.0",
    "eslint-plugin-jest": "^22.21.0",
    "eslint-plugin-prettier": "^3.1.3",
    "eslint-plugin-typescript": "^0.14.0",
    "jest": "^26.2.2",
    "nock": "^11.9.1",
    "prettier": "^2.0.5",
    "smee-client": "^1.1.0",
    "ts-jest": "^26.1.4",
    "typescript": "^3.9.7"
  },
  "engines": {
    "node": ">=12.8.1"
  }
}
