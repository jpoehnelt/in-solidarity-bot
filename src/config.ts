/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DEFAULT_RULES, Rule } from "./rules";

import Ajv from "ajv";
import { Context } from "probot";
import { Level } from "./rules";
import deepmerge from "deepmerge";

export interface Configuration {
  rules: { [key: string]: Rule };
  ignore: string[];
}
export class InvalidConfigError extends Error {}

export const DEFAULT_CONFIGURATION: Configuration = {
  rules: DEFAULT_RULES,
  ignore: [".github/in-solidarity.yml"],
};

const CONFIG_FILE = "in-solidarity.yml";

const ajv = new Ajv({ allErrors: true });

const rulesPropertiesSchema = Object.keys(DEFAULT_RULES).reduce((obj, k) => {
  obj[k] = {
    type: "object",
    additionalProperties: false,
    properties: {
      level: { type: "string", enum: Object.values(Level) },
    },
  };
  return obj;
}, {});

const schema = {
  type: "object",
  additionalProperties: false,
  properties: {
    rules: {
      type: "object",
      additionalProperties: false,
      properties: rulesPropertiesSchema,
    },
    ignore: {
      type: "array",
      items: { type: "string" },
    },
  },
};

export const getConfig = async (context: Context): Promise<Configuration> => {
  const validate = ajv.compile(schema);

  const repoConfig = await context.config(CONFIG_FILE);
  if (repoConfig) {
    if (!validate(repoConfig)) {
      throw new InvalidConfigError(
        "configuration is invalid: " + JSON.stringify(validate.errors)
      );
    }

    return deepmerge(DEFAULT_CONFIGURATION, repoConfig as Configuration, {
      arrayMerge: (_, b) => b,
    });
  }
  return DEFAULT_CONFIGURATION;
};
