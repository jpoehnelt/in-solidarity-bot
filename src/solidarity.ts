/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Configuration, InvalidConfigError, getConfig } from "./config";
import { annotate, getLevelFromAnnotations } from "./annotate";

import { Context } from "probot";
import { File } from "gitdiff-parser";
import { Level } from "./rules";
import { LoggerWithTarget } from "probot/lib/wrap-logger";
import { Octokit } from "@octokit/rest/";
import fs from "fs";
import { parse } from "./parse";

const SUMMARY = fs.readFileSync("./static/HELP.md", "utf8");
const CHECK_NAME = "Inclusive Language";

export enum Conclusion {
  SUCCESS = "success",
  FAILURE = "failure",
  NEUTRAL = "neutral",
  CANCELLED = "cancelled",
  TIMED_OUT = "timed_out",
  ACTION_REQUIRED = "action_required",
}

export enum OutputTitle {
  SUCCESS = "Check completed with success",
  ERROR = "Check failed due to error",
  PERMISSION_NEEDED = "Check lacks permissions for private repository",
  FAILURE = "Check completed with failures",
  NOTICE = "Check completed with notices",
  WARNING = "Check completed with warnings",
}

export class Solidarity {
  private context: Context;
  private logger: LoggerWithTarget;
  private checkId?: number;
  config?: Configuration;

  constructor(context: Context, logger: LoggerWithTarget) {
    this.context = context;
    this.logger = logger;
  }

  get headSha(): string {
    return this.context.payload.pull_request.head.sha;
  }

  get owner(): string {
    return this.context.payload.repository.owner.login;
  }

  get repo(): string {
    return this.context.payload.repository.name;
  }

  get pullNumber(): number {
    return this.context.payload.number;
  }

  get checkOptions(): Octokit.ChecksCreateParams {
    return {
      owner: this.owner,
      repo: this.repo,
      head_sha: this.headSha,
      name: CHECK_NAME,
    };
  }

  async run(): Promise<void> {
    let conclusion: Conclusion = Conclusion.NEUTRAL;
    let output: Octokit.ChecksUpdateParamsOutput;

    await this.start();
    await this.update("in_progress");

    try {
      this.config = await getConfig(this.context);
    } catch (e) {
      if (e instanceof InvalidConfigError) {
        conclusion = Conclusion.FAILURE;
        output = {
          title: OutputTitle.ERROR,
          summary: "Configuration is invalid.",
        };
      } else {
        conclusion = Conclusion.CANCELLED;
        output = {
          title: OutputTitle.ERROR,
          summary: "Could not load configuration.",
        };
      }

      this.logger.error(
        { labels: { ...this.checkOptions }, err: e },
        output.summary
      );

      await this.update("completed", conclusion, output);
      return;
    }

    try {
      const check = await this.check();
      conclusion = check.conclusion;
      output = check.output;
    } catch (e) {
      if (e.status === 403) {
        output = {
          title: OutputTitle.PERMISSION_NEEDED,
          summary:
            "Check only runs on public repositories to limit required permissions. See https://github.com/jpoehnelt/in-solidarity-bot/issues/16.",
        };

        this.logger.info(
          { labels: { ...this.checkOptions }, err: e },
          "Failed to check private repository"
        );
      } else {
        output = {
          title: OutputTitle.ERROR,
          summary: "Check failed to complete.",
        };

        this.logger.error(
          { labels: { ...this.checkOptions }, err: e },
          "Failed to complete check"
        );
      }
    }

    await this.update("completed", conclusion, output);
  }

  async start(): Promise<void> {
    try {
      const response = await this.context.github.checks.create({
        ...this.checkOptions,
        status: "queued",
      });
      this.checkId = response.data.id;
    } catch (e) {
      this.logger.error({ err: e }, "Failed to create check");
    }
  }

  async update(
    status: "queued" | "in_progress" | "completed",
    conclusion?: Conclusion,
    output?: Octokit.ChecksUpdateParamsOutput,
    details_url?: string
  ): Promise<void> {
    try {
      await this.context.github.checks.update({
        ...this.checkOptions,
        check_run_id: this.checkId as number,
        status,
        conclusion,
        ...(status === "completed" && {
          completed_at: new Date().toISOString(),
        }),
        ...(output && { output }),
        details_url,
      });
    } catch (e) {
      this.logger.error({ err: e }, "Failed to update check");
    }
  }

  async diff(): Promise<File[]> {
    const response = await this.context.github.pulls.get({
      owner: this.owner,
      repo: this.repo,
      pull_number: this.pullNumber,
      headers: { accept: "application/vnd.github.v3.diff" },
    });

    const diff = (response.data as unknown) as string;

    return parse(diff);
  }

  async check(): Promise<{
    conclusion: Conclusion;
    output: Octokit.ChecksUpdateParamsOutput;
  }> {
    let conclusion: Conclusion;
    const output: Octokit.ChecksUpdateParamsOutput = {
      title: CHECK_NAME,
      summary: SUMMARY,
    };
    const diff = await this.diff();

    output.annotations = annotate(this.config as Configuration, diff);

    const level = getLevelFromAnnotations(output.annotations);

    switch (level) {
      case Level.FAILURE:
        conclusion = Conclusion.ACTION_REQUIRED;
        output.title = OutputTitle.FAILURE;
        break;
      case Level.WARNING:
        conclusion = Conclusion.NEUTRAL;
        output.title = OutputTitle.WARNING;
        break;
      case Level.NOTICE:
        conclusion = Conclusion.NEUTRAL;
        output.title = OutputTitle.NOTICE;
        break;
      case Level.OFF:
      default:
        conclusion = Conclusion.SUCCESS;
        output.title = OutputTitle.SUCCESS;
    }

    this.logger.info({
      conclusion,
      repo: this.repo,
      owner: this.owner,
      pull_number: this.pullNumber,
      sha: this.headSha,
    });

    return { conclusion, output };
  }
}
