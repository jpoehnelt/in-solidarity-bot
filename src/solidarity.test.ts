/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Conclusion, OutputTitle, Solidarity, logger } from "./solidarity";

import { DEFAULT_CONFIGURATION } from "./config";
import fs from "fs";
import { parse } from "./parse";
import { version } from "../package.json";

const payload = JSON.parse(fs.readFileSync("./fixtures/payload.json", "utf8"));
const failingDIff = parse(
  fs.readFileSync("./fixtures/pull.failing.diff", "utf8")
);

beforeAll(() => {
  jest.mock("../package.json", () => {
    return {
      version: "1.1.0",
    };
  });
});

test("solidarity should instantiate child logger", () => {
  const child = jest.spyOn(logger, "child");

  const s = new Solidarity({ name: "foo", id: "bar", payload: payload } as any);
  expect(child).toHaveBeenCalledWith({
    owner: "jpoehnelt",
    pull_number: 24,
    repo: "in-solidarity-bot",
    sha: "d405f6ab463fcc7dc6717bd706d0529d5a71ba53",
    version,
  });
});

test("solidarity should run check on failing diff", async () => {
  const s = new Solidarity({ name: "foo", id: "bar", payload: payload } as any);

  s.diff = jest.fn().mockReturnValue(failingDIff);
  s.config = DEFAULT_CONFIGURATION;

  const info = jest.spyOn(s["logger"], "info").mockImplementation((msg) => msg);
  const { conclusion, output } = await s.check();
  expect(conclusion).toBe(Conclusion.NEUTRAL);
  expect(output.title).toEqual(OutputTitle.WARNING);
  expect(output.annotations?.length).toEqual(2);
  expect(info.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "column": 0,
          "content": "whitelist",
          "level": "warning",
          "line": 2,
          "path": "README.md",
          "regex": "/white[_-]*list/gi",
        },
        "match",
      ],
      Array [
        Object {
          "column": 0,
          "content": "Master",
          "level": "warning",
          "line": 3,
          "path": "README.md",
          "regex": "/master/gi",
        },
        "match",
      ],
      Array [
        Object {
          "conclusion": "neutral",
        },
      ],
    ]
  `);
});

test("solidarity should have correct properties from payload", async () => {
  const s = new Solidarity({ name: "foo", id: "bar", payload: payload } as any);
  expect(s.owner).toEqual("jpoehnelt");
  expect(s.repo).toEqual("in-solidarity-bot");
  expect(s.headSha).toEqual("d405f6ab463fcc7dc6717bd706d0529d5a71ba53");
  expect(s.checkOptions).toEqual({
    owner: "jpoehnelt",
    repo: "in-solidarity-bot",
    head_sha: "d405f6ab463fcc7dc6717bd706d0529d5a71ba53",
    name: "Inclusive Language",
  });
  expect(s.pullNumber).toEqual(24);
});

test("solidarity should generate correct summary", async () => {
  const s = new Solidarity({ name: "foo", id: "bar", payload: payload } as any);
  s.config = DEFAULT_CONFIGURATION;
  expect(s.summary("foo", "1.0.0")).toMatchSnapshot();
});
