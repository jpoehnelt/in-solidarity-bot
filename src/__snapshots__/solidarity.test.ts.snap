// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`solidarity should generate correct summary 1`] = `
"## Message
foo

## Configuration
Below is the configuration used for this check run:
\`\`\`yml
rules:
  master:
    regex:
      - /master/gi
    level: warning
    alternatives:
      - primary
      - main
      - leader
      - active
      - writer
  slave:
    regex:
      - /slave/gi
    level: warning
  whitelist:
    regex:
      - '/white[_-]*list/gi'
    level: warning
  blacklist:
    regex:
      - '/black[_-]*list/gi'
    level: warning
  grandfathered:
    regex:
      - /grandfathered/gi
    level: warning
  sanity_check:
    regex:
      - '/sanity[_-]*check/gi'
    level: warning
  man_hours:
    regex:
      - '/man[_-]*hours/gi'
    level: warning
ignore:
  - .github/in-solidarity.yml

\`\`\`

For more information on configuration and rules, check the [documentation](https://github.com/jpoehnelt/in-solidarity-bot/blob/main/docs/README.md).

## App version
* version: [v1.0.0](https://github.com/jpoehnelt/in-solidarity-bot/releases/tag/v1.0.0)

* sha: [unknown](https://github.com/jpoehnelt/in-solidarity-bot/commit/unknown)
"
`;

exports[`solidarity should run check on failing diff 1`] = `
"
## Configuration
Below is the configuration used for this check run:
\`\`\`yml
rules:
  master:
    regex:
      - /master/gi
    level: warning
    alternatives:
      - primary
      - main
      - leader
      - active
      - writer
  slave:
    regex:
      - /slave/gi
    level: warning
  whitelist:
    regex:
      - '/white[_-]*list/gi'
    level: warning
  blacklist:
    regex:
      - '/black[_-]*list/gi'
    level: warning
  grandfathered:
    regex:
      - /grandfathered/gi
    level: warning
  sanity_check:
    regex:
      - '/sanity[_-]*check/gi'
    level: warning
  man_hours:
    regex:
      - '/man[_-]*hours/gi'
    level: warning
ignore:
  - .github/in-solidarity.yml

\`\`\`

For more information on configuration and rules, check the [documentation](https://github.com/jpoehnelt/in-solidarity-bot/blob/main/docs/README.md).

## App version
* version: [v1.0.0](https://github.com/jpoehnelt/in-solidarity-bot/releases/tag/v1.0.0)

* sha: [unknown](https://github.com/jpoehnelt/in-solidarity-bot/commit/unknown)
"
`;
